bibliopixel/drivers/SPI/APA102.py

```diff
@@ -1,24 +1,43 @@
 from ... import gamma
 from .. channel_order import ChannelOrder
+from ... import data_maker
 from . base import SPIBase
 
 
 class APA102(SPIBase):
-    """Driver for APA102 based LED strips on devices like the Raspberry Pi and BeagleBone"""
+    """Driver for APA102/SK9822 based LED strips on devices like the Raspberry Pi and BeagleBone"""
 
     def __init__(self, num, gamma=gamma.APA102, **kwargs):
         super().__init__(num, gamma=gamma, **kwargs)
+        maker = kwargs.get('maker', data_maker.MAKER)
 
-        # APA102 requires latch bytes at the end
-        self._latch_bytes = [0xFF, 0, 0, 0] * ((num // 64) + 1)
+        # APA102/SK9822 requires latch bytes at the end)
+        # Many thanks to this article for combined APA102/SK9822 protocol
+        # https://cpldcpu.com/2016/12/13/sk9822-a-clone-of-the-apa102/
+        self._start_frame = 4  # start frame is [0, 0, 0, 0]
+        self._pixel_bytes = self.bufByteCount() + self.numLEDs  # 1 extra for brightness
+        self._pixel_stop = self._start_frame + self._pixel_bytes
+        self._reset_frame = 4  # for SK9822 [0, 0, 0, 0]
+        self._end_frame = ((num // 2) + 1)
+        self._packet = maker.make_packet(self._start_frame + self._pixel_bytes +
+                                         self._reset_frame + self._end_frame)
+
+        self.set_device_brightness(0xFF)  # required to setup _packet
+
+    def set_device_brightness(self, val):
+        """
+        APA102 & SK9822 support on chip brightness control allowing greater color depth.
+        APA102 superimposes a 440Hz PWM on the 19kHz base PWM to control brightness.
+        SK9822 uses a base 4.7kHz PWM but controls brightness with a variable current source.
+        Because of this SK9822 will have much less flicker at lower levels.
+        Either way, this option is better and faster than scaling in BiblioPixel
+        """
+        self._chipset_brightness = (val >> 3)  # bitshift to scale from 8 bit to 5
+        self._brightness_list = [0xE0 + self._chipset_brightness] * self.numLEDs
```

Comment:
Read this: https://cpldcpu.com/2016/12/13/sk9822-a-clone-of-the-apa102/
Basically, the first byte of the 4 byte data frame is brightness, but they only support 5-bit brightness control. Every other driver for APA102 you will see just sets that to 0xFF (full brightness) and then handles brightness in software (yes, we are the first to FULLY support this chipset!). So, per the protocol spec, 0xE0 is 0% brightness and then you can add up to 31 to that to get full brightness. This is why I just take the 8-bit value we normally use and do the `>> 3` to it, to scale it to 5-bits.